File: app/src/main/java/mhashim6/android/quickQuery/ClipboardMonitor.java
package mhashim6.android.quickQuery;

import android.app.Service;
import android.content.ClipboardManager;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.PixelFormat;
import android.os.IBinder;
import android.preference.PreferenceManager;
import android.util.Log;
import android.view.Gravity;
import android.view.View;
import android.view.WindowManager;
import android.widget.ImageView;

import java.net.MalformedURLException;
import java.net.URL;

import androidx.annotation.Nullable;

import static android.content.Intent.FLAG_ACTIVITY_NEW_TASK;
import static mhashim6.android.quickQuery.NotificationsManagerKt.buildForegroundNotification;
import static mhashim6.android.quickQuery.Utils.IS_OREO;
import static mhashim6.android.quickQuery.Utils.QUICK_QUERY_ACTION;


/**
 * Created by mhashim6 on 25/10/2017.
 */

public class ClipboardMonitor extends Service {

	private static final int FOREGROUND_ID = 77;

	private ImageView bubble;
	private static final WindowManager.LayoutParams LAYOUT_PARAMS = new WindowManager.LayoutParams(
			WindowManager.LayoutParams.WRAP_CONTENT,
			WindowManager.LayoutParams.WRAP_CONTENT,
			IS_OREO ?
					WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
					: WindowManager.LayoutParams.TYPE_PHONE,
			WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
					| WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN,
			PixelFormat.TRANSLUCENT);

	static {
		LAYOUT_PARAMS.gravity = Gravity.TOP | Gravity.START;
		LAYOUT_PARAMS.x = 3;
		LAYOUT_PARAMS.y = 180;
	}

	private CharSequence query;
//===================================================

	@Override
	public void onCreate() {

		super.onCreate();

		if (IS_OREO)
			startForeground(FOREGROUND_ID, buildForegroundNotification(this));
		else
			Log.i("QuickQuery", "not oreo, skipping notification.");

		initBubble();

		ClipboardManager clipboardManager = (ClipboardManager) getSystemService(CLIPBOARD_SERVICE);
		clipboardManager.addPrimaryClipChangedListener(() -> handleNewQuery(clipboardManager.getPrimaryClip().getItemAt(0).getText()));

		/*
		if (BuildConfig.FLAVOR.equals(FLAVOR_FULL))
			MobileAds.initialize(this, ADMOB_APP_ID);
		*/
	}
//===================================================

	private void initBubble() {
		this.bubble = new ImageView(this);
		bubble.setImageResource(R.drawable.ic_bubble);
	}

	private void initAndAddNewBubbleToWindow() {
		initBubble();

		WindowManager windowManager = (WindowManager) getSystemService(WINDOW_SERVICE);
		windowManager.addView(bubble, LAYOUT_PARAMS);
	}

	private void handleNewQuery(CharSequence query) {
		try {
			new URL(query.toString()); //ignore if url.
		} catch (MalformedURLException e) {
			this.query = query;
			showNewBubble();
		}
	}

	private void showNewBubble() {
		bubble.setVisibility(View.GONE);

		initAndAddNewBubbleToWindow();
		View newBubble = this.bubble;

		newBubble.setOnClickListener(view -> {
			newBubble.setVisibility(View.GONE);
			startQQActivity(query);
		});
		newBubble.setOnLongClickListener(view -> {//hide bubble
			newBubble.setVisibility(View.GONE);
			return true;
		});

		SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(this);
		long duration = Long.parseLong(preferences.getString("duration", "6")) * 1000;
		newBubble.postDelayed(() -> removeBubbleFromWindow(newBubble), duration);
	}

	private void removeBubbleFromWindow(View bubble) {
		WindowManager windowManager = (WindowManager) getSystemService(WINDOW_SERVICE);
		windowManager.removeView(bubble);
	}
//===================================================

	private void startQQActivity(CharSequence query) {
		Intent qqStarter = new Intent(this, QQActivity.class);
		qqStarter.setAction(QUICK_QUERY_ACTION);
		qqStarter.putExtra(Intent.EXTRA_TEXT, query);
		qqStarter.setFlags(FLAG_ACTIVITY_NEW_TASK);
		startActivity(qqStarter);
	}
//===================================================

	@Nullable
	@Override
	public IBinder onBind(Intent intent) {
		return null;
	}

}

File: app/src/main/java/mhashim6/android/quickQuery/ClipboardMonitor.kt
package mhashim6.android.quickQuery

import android.annotation.SuppressLint
import android.app.Service
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import android.content.Intent.FLAG_ACTIVITY_NEW_TASK
import android.graphics.PixelFormat
import android.os.IBinder
import android.preference.PreferenceManager
import android.util.Log
import android.view.Gravity
import android.view.View
import android.view.WindowManager
import android.widget.ImageView
import kotlinx.coroutines.launch
import java.net.MalformedURLException
import java.net.URL


/**
 * Created by mhashim6 on 25/10/2017.
 */

class ClipboardMonitor : Service() {

    private lateinit var bubble: ImageView

    private var query: CharSequence? = null

    @SuppressLint("NewApi")
    override fun onCreate() {

        super.onCreate()

        if (IS_OREO)
            startForeground(FOREGROUND_ID, buildForegroundNotification(this))
        else
            Log.i("QuickQuery", "not oreo, skipping notification.")

        initBubble()

        val clipboardManager = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        clipboardManager.addPrimaryClipChangedListener { handleNewQuery(clipboardManager.primaryClip?.getItemAt(0)?.text) }
    }

    private fun initBubble() {
        bubble = ImageView(this)
        bubble.setImageResource(R.drawable.ic_bubble)
    }

    private fun initAndAddNewBubbleToWindow() {
        initBubble()

        val windowManager = getSystemService(Context.WINDOW_SERVICE) as WindowManager
        windowManager.addView(bubble, LAYOUT_PARAMS)
    }

    private fun handleNewQuery(query: CharSequence?) {
        query?.let {
            try {
                URL(it.toString()) //ignore if url.
            } catch (e: MalformedURLException) {
                this.query = it
                mainScope.launch {
                    showNewBubble()
                }
            }
        }
    }

    private fun showNewBubble() {
        bubble.visibility = View.GONE

        initAndAddNewBubbleToWindow()
        val newBubble = this.bubble

        newBubble.setOnClickListener {
            newBubble.visibility = View.GONE
            startQQActivity(query)
        }
        newBubble.setOnLongClickListener {
            newBubble.visibility = View.GONE
            true
        }

        val preferences = PreferenceManager.getDefaultSharedPreferences(this)
        val duration = (preferences.getString("duration", "6")).toLong() * 1000
        newBubble.postDelayed({ removeBubbleFromWindow(newBubble) }, duration)
    }

    private fun removeBubbleFromWindow(bubble: View) {
        val windowManager = getSystemService(Context.WINDOW_SERVICE) as WindowManager
        windowManager.removeView(bubble)
    }

    private fun startQQActivity(query: CharSequence?) {
        val qqStarter = Intent(this, QQActivity::class.java)
        qqStarter.action = QUICK_QUERY_ACTION
        qqStarter.putExtra(Intent.EXTRA_TEXT, query)
        qqStarter.flags = FLAG_ACTIVITY_NEW_TASK
        startActivity(qqStarter)
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    companion object {
        private const val FOREGROUND_ID = 77
        private val LAYOUT_PARAMS = WindowManager.LayoutParams(
                WindowManager.LayoutParams.WRAP_CONTENT,
                WindowManager.LayoutParams.WRAP_CONTENT,
                if (IS_OREO)
                    WindowManager.LayoutParams.TYPE_APPLICATION_OVERLAY
                else
                    WindowManager.LayoutParams.TYPE_PHONE,
                WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE or WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN,
                PixelFormat.TRANSLUCENT)

        init {
            LAYOUT_PARAMS.gravity = Gravity.TOP or Gravity.START
            LAYOUT_PARAMS.x = 3
            LAYOUT_PARAMS.y = 180
        }
    }

}
